sdk/src/types/block/address/restricted.rs

```diff
@@ -0,0 +1,204 @@
+        let address = address.into();
+        if matches!(address, Address::Restricted(_)) {
+            return Err(Error::InvalidAddressKind(Self::KIND));
+        }
+        Ok(Self {
+            address,
+            allowed_capabilities: Default::default(),
+        })
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn with_allowed_capabilities(
+        mut self,
+        allowed_capabilities: impl IntoIterator<Item = impl Into<Capabilities>>,
+    ) -> Result<Self, Error> {
+        self.allowed_capabilities = allowed_capabilities
+            .into_iter()
+            .map(Into::into)
+            .collect::<Box<[_]>>()
+            .try_into()
+            .map_err(Error::InvalidCapabilitiesCount)?;
+        Ok(self)
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn set_allowed_capabilities(
+        &mut self,
+        allowed_capabilities: impl IntoIterator<Item = impl Into<Capabilities>>,
+    ) -> Result<&mut Self, Error> {
+        self.allowed_capabilities = allowed_capabilities
+            .into_iter()
+            .map(Into::into)
+            .collect::<Box<[_]>>()
+            .try_into()
+            .map_err(Error::InvalidCapabilitiesCount)?;
+        Ok(self)
+    }
+
+    /// Gets the allowed capabilities.
+    pub fn allowed_capabilities(&self) -> &[Capabilities] {
+        &self.allowed_capabilities
+    }
+}
+
+impl TryFrom<Address> for RestrictedAddress {
+    type Error = Error;
+
+    fn try_from(value: Address) -> Result<Self, Self::Error> {
+        Self::new(value)
+    }
+}
+
+impl core::fmt::Display for RestrictedAddress {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", prefix_hex::encode(self.pack_to_vec()))
+    }
+}
+
+impl Packable for RestrictedAddress {
+    type UnpackError = Error;
+    type UnpackVisitor = ();
+
+    fn pack<P: packable::packer::Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.address.pack(packer)?;
+        if self.allowed_capabilities.iter().any(|c| c.0 != CapabilityFlag::NONE) {
+            self.allowed_capabilities.pack(packer)?;
+        } else {
+            0_u8.pack(packer)?;
+        }
+        Ok(())
+    }
+
+    fn unpack<U: packable::unpacker::Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        _visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, packable::error::UnpackError<Self::UnpackError, U::Error>> {
+        let address = Address::unpack::<_, VERIFY>(unpacker, &())?;
+        let allowed_capabilities = BoxedSlicePrefix::<_, CapabilitiesCount>::unpack::<_, VERIFY>(unpacker, &())
+            .map_packable_err(|e| Error::InvalidCapabilitiesCount(e.into_prefix_err().into()))
+            .coerce()?;
+        Ok(Self {
+            address,
+            allowed_capabilities,
+        })
+    }
+}
+
+pub struct CapabilityFlag;
+
+impl CapabilityFlag {
```

Comment:
why not an enum? So that we cannot just add variants i guess? Could always make a custom variant

Reply:
The main reason I didn't go with an enum is that it can't be used like this:
```rust
if capabilities.has_capabilities(Flag::NativeTokens | Flag::Mana) {...}

if capabilities.has_capabilities(Flag::All) {...}
```

Reply:
good point! could impl deref but not too nice

Reply:
Even deref unfortunately would not allow this kind of usage, you have to add `as u8` to each item